src/PrintProvider.jsx

```diff
@@ -0,0 +1,102 @@
+import React from 'react';
+import ReactDOM from 'react-dom';
+import PropTypes from 'prop-types';
+import s from './styles.css';
+import { spliced, debug } from './common';
+
+export const propTypes = {
+  loose: PropTypes.bool,
+  children: PropTypes.node.isRequired,
+  invert: PropTypes.bool,
+};
+
+export const childContextTypes = {
+  printProvider: PropTypes.shape({
+    regPrintable: PropTypes.func.isRequired,
+    unregPrintable: PropTypes.func.isRequired,
+    isPrint: PropTypes.bool.isRequired,
+  }).isRequired,
+};
+
+export default class PrintProvider extends React.PureComponent {
+  constructor(props) {
+    super(props);
+    this.state = {
+      isInPrintPreview: false,
+      printableNodes: [],
+      hasSingle: false,
+    };
+    this.printableRegistry = {};
+    window.matchMedia('print').onchange = () => {
+      debug('toggle print mode', window.matchMedia('print').matches);
+      this.setState({ isInPrintPreview: window.matchMedia('print').matches });
+    };
+  }
+
+  getChildContext() {
+    return {
+      printProvider: {
+        isPrint: window.matchMedia('print').matches,
+        regPrintable: this.regPrintable.bind(this),
+        unregPrintable: this.unregPrintable.bind(this),
+      },
+    };
+  }
+
+  regPrintable(key, node, isSingle) {
+    const loose = this.props.loose || this.props.invert;
+    debug('reg printable', key, node);
+    if (this.printableRegistry[key] !== undefined || loose) return;
+    setTimeout(() => this.setState({
+      printableNodes: this.state.printableNodes.concat(node),
+      hasSingle: this.state.hasSingle || isSingle
+    }), 0);
+    this.printableRegistry[key] = this.state.printableNodes.length;
+  }
+
+  unregPrintable(key) {
+    const loose = this.props.loose || this.props.invert;
+    if (this.printableRegistry[key] === undefined || this.state.isInPrintPreview || loose) return;
+    this.setState({
+      printableNodes: spliced(this.state.printableNodes, this.printableRegistry[key]),
+    });
+    this.printableRegistry = Object.assign({}, this.printableRegistry, { [key]: undefined });
+  }
+
+  createRender(children) {
+    const el = document.createElement('div');
+    el.id = 'render';
+    document.body.appendChild(el);
+
+    ReactDOM.render(<div className={s.printRender} >{children}</div>, el);
+  }
+
+  deleteRender() {
+    const el = document.getElementById('render');
+    el && document.body.removeChild(el);
+  }
+
+  render() {
+    const { isInPrintPreview, printableNodes, hasSingle } = this.state;
+    const loose = this.props.loose || this.props.invert;
+
+    if (isInPrintPreview && printableNodes.length && !loose && !hasSingle) {
+      debug('render printable only', printableNodes);
+      const children = React.Children.map(printableNodes, (child,key) => {
+        return React.cloneElement(child, { key });
+      });
+
+      this.createRender(<div>{children}</div>);
```

Comment:
The offset (you used above) is good enough solution. We're don't need a React.Portal-like thing, for now, isn't it?

Reply:
Offset thing is being used only for (single, main) Print component, we want to use (single, main)  prop when we don't want to lose the state of Print component's children and we only want to print only one particular component.

In other cases (when we want to print more than one Print component) the lib will create an extra DOM node and will render the Print components there

Reply:
We can print one particular component only using offsets and css visibility hacks w/o re-rendering anymore, am I correct?

Reply:
@a-x- yes, and if there are several of Print components, we're re-rendering them in the new DOM node, when print mode is on, then deleting this DOM node.